### PR TITLE
Display X/Twitter API rate limits

### PR DESCRIPTION
**Pull request by s3rgiosan**
### Description of the Change

Display X/Twitter API rate limits 

### Changelog Entry

> Added - Admin dashboard widget for easy access to API usage statistics.
> Changed - Display API usage statistics in the "Autopost to X/Twitter" block editor panel.

### Credits

Props @s3rgiosan

### Checklist:

- [ ] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [ ] I have updated the documentation accordingly.
- [ ] I have added [Critical Flows, Test Cases, and/or End-to-End Tests](https://10up.github.io/Open-Source-Best-Practices/testing/) to cover my change.
- [ ] All new and existing tests pass.

### CODE REVIEW

**Comment by s3rgiosan** [comment]
Dashboard widget
Screenshot 2024-12-16 at 17 07 19

Block editor
Screenshot 2024-12-16 at 17 07 15

File: includes/core.php [hunk 1]
```diff
@@ -222,3 +224,225 @@ function handle_notice_dismiss() {
...
+					);
+				},
+				$rows
+			);
+
+			$account_markup = implode( ' ', $account_markup );
+		} else {
+			$account_markup = sprintf(
+				'<p>%s</p>',
+				esc_html__( 'No X/Twitter rate data available yet. Make a post to X/Twitter first.', 'autoshare-for-twitter' )
```
Comment (iamdharmesh) [inline includes/core.php]:
```suggestion
				esc_html__( 'No X/Twitter rate limit data available yet. Make a post to X/Twitter first.', 'autoshare-for-twitter' )
```

**Review by iamdharmesh — COMMENTED** [review]
Thanks a lot for the PR, @s3rgiosan! This looks great and tests well. I just have a few points to discuss before we conclude and mark this as ready for merge:

1. The rate limit for the 15-minute window is very high (1079999 out of 1080000). I don't think it will ever be utilized. How about we drop displaying it and only show the 24-hour limit for the user and app?

2. In the Block editor panel, the note under the limits currently appears for each account. Maybe we should display it only once, similar to how it's done for the dashboard widget. Additionally, how about keeping the rate limits data hidden until the account toggle is enabled?

Screenshot 2024-12-17 at 4 04 16 PM

Thank you.

File: includes/core.php (hunk 1 above)
Comment (s3rgiosan) [reply includes/core.php]: Updated

**Comment by s3rgiosan** [comment]
Hi @iamdharmesh 

> The rate limit for the 15-minute window is very high (1079999 out of 1080000). I don't think it will ever be utilized. How about we drop displaying it and only show the 24-hour limit for the user and app?

Agreed. Since those headers are already being stored, they'll be available if we ever need to display them. I've removed them from both the dashboard and the block editor panel.

> In the Block editor panel, the note under the limits currently appears for each account. Maybe we should display it only once, similar to how it's done for the dashboard widget. Additionally, how about keeping the rate limits data hidden until the account toggle is enabled?

Both things were addressed. Can you review again?

Thanks!

**Comment by iamdharmesh** [comment]
Hi @s3rgiosan, 

Thanks for the quick updates on this; it looks great now. Apologies for bringing up one more thing to discuss. As we are using the same app for all accounts, it would be great if the limits could be updated for each account whenever a tweet is posted from any account. Currently, it only reflects on the account from which the tweet was made. This is not a major issue and only comes into play when multiple accounts are connected to the site (which is not always the case), but I just wanted to hear your thoughts on this. Otherwise, everything looks good to me.

Screenshot 2024-12-17 at 6 04 05 PM
 
 Thank you.

**Comment by s3rgiosan** [comment]
@iamdharmesh Does the plugin support multiple connected apps, or is it limited to a single one? If it's limited to one, we could store the app's rate limits at a global level instead of per account. Alternatively, whenever a tweet is made, we could iterate through all accounts to update their corresponding app limit values.

**Comment by iamdharmesh** [comment]
> @iamdharmesh Does the plugin support multiple connected apps, or is it limited to a single one?

@s3rgiosan it is limited to a single app. All user accounts are connected using the same app.

**Comment by s3rgiosan** [comment]
Thanks. I'll take a look and update the PR

**Comment by s3rgiosan** [comment]
Hi @iamdharmesh — I’ve updated the PR as requested. The limits are now displayed per user account, followed by the app limits. The app limits are fetched from the first user account. 
When a tweet is sent, only the user account that sent the tweet gets its user-specific limits updated, and all user accounts are updated for "global" and app limits.

Dashboard
Screenshot 2024-12-17 at 23 34 42

With one account
Screenshot 2024-12-17 at 23 42 09


With more than one account
Screenshot 2024-12-17 at 23 41 50

**Review by iamdharmesh — APPROVED** [review]
@s3rgiosan Thanks for making all the changes here! Things look good, test well, and are ready to be merged.

_Note: I made some minor [changes](https://github.com/10up/autoshare-for-twitter/pull/354/commits/cd51aa2d9cfed07dbd2679163cab2522958b1220) to fix the user rate limit reset issue for another user and updated the UI slightly_

**Pull request merged**